MySQL single master, multi-slave

jpoon opened the pull request:
This chart builds upon this [example](https://kubernetes.io/docs/tutorials/stateful-application/run-replicated-stateful-application/) to deploy a MySQL cluster with a single master that is replicated across multiple slaves.  

Tested with MySQL 5.7.13

jpoon committed:
add mysql master-slave

jpoon committed:
update mysqlha readme

viglesiasce commented:
@jpoon could this be implemented as an extension to the existing mysql chart? Configurable through the values file?

viglesiasce commented on incubator/mysqlha/templates/statefulset.yaml:
```diff
@@ -0,0 +1,249 @@
+            "name": "init-mysql",
+            "image": "mysql:{{ .Values.imageTag }}",
+            "command": ["bash", "-c", "
+              set -ex\n
```
Since there is a lot going on in here can you split this out into its own file then use a Files.Get+ indent to add it in here?
There is an example in the spinnaker chart:
https://github.com/kubernetes/charts/blob/master/stable/spinnaker/templates/configmap/spinnaker-config.yaml#L11

viglesiasce commented on incubator/mysqlha/templates/statefulset.yaml:
```diff
@@ -0,0 +1,249 @@
+          initialDelaySeconds: 10
+          timeoutSeconds: 1
+      - name: xtrabackup
+        image: gcr.io/google-samples/xtrabackup:1.0
```
Make this configurable

viglesiasce commented on incubator/mysqlha/templates/statefulset.yaml:
```diff
@@ -0,0 +1,249 @@
+        - |
+          set -ex
+
+          echo "Waiting for mysqld to be ready (accepting connections)"
```
Same as above. Can you put this in its own file and then bring it in with Files.Get

jpoon replied:
Spent an entire morning and couldn't get this to work with `Files.Get`.

escardin replied:
I think this is now resolved

viglesiasce commented on incubator/mysqlha/values.yaml:
```diff
@@ -0,0 +1,42 @@
+  ## If defined, volume.beta.kubernetes.io/storage-class: <storageClass>
+  ## Default: volume.alpha.kubernetes.io/storage-class: default
+  ##
+  storageClass: fast 
```
Do not set this by default in order to allow the default storage class of the cluster to be used.

viglesiasce commented on incubator/mysqlha/templates/statefulset.yaml:
```diff
@@ -0,0 +1,249 @@
+    spec:
+      containers:
+      - name: mysql
+        image: mysql:{{ .Values.imageTag }}
```
allow image repository to also be configurable

viglesiasce reviewed (changes requested):
Was able to get the master to come up but then it kept crashing. The 2nd pod was stuck in a crash backoff loop:

```
mysql-ha-mysqlha-0                                   2/2       Running                 2          4m
mysql-ha-mysqlha-1                                   0/2       Init:CrashLoopBackOff   4          3m
```

jpoon commented:
Thanks for the review @viglesiasce. I'm a lil overwhelmed at work at the moment and will address them as soon as I can.

prydonius commented:
@jpoon gentle ping, are you able to pick this up again?

jpoon commented:
@prydonius thanks for the reminder. been busy at work; hope to pick this back up and update the PR by mid-June.

prydonius commented:
@jpoon no problem, since this is for the incubator, we could merge as-is so people can try it out/improve in the meantime. @viglesiasce we can visit the idea of merging this with stable/mysql when we move this to stable, would you be okay with merging as-is for now?

prydonius commented:
@jpoon I think since the second pod is crashing we shouldn't merge this right now. Do you think you'll be able to pick this up now? If not I suggest we close this and you can always resubmit when you have the time

jpoon commented:
Looks like this chart broke with k8s 1.6+, and the `init-containers` also went out of `beta` so there's some work to refactor the code. I'll close this PR and re-submit later.

Commit:
Refactored initContainers

Commit:
Changed storageClass to default

Commit:
Made mysqlImage and xtraBackupImage configurable

Commit:
Refactored initContainers to use Files.Get

Commit:
Removed default storageClass as not required

Not specifying a storageClass will cause k8s to use the default
storageClass anyway.

jpoon committed:
update readme. add jon as maintainer :D

jpoon commented:
Thanks to @jonhoare for getting this work! Re-opening this PR.

jonhoare commented:
CLA signed.

jpoon commented:
Weird, the bot is still complaining. @jonhoare did you sign the CLA with the same email as the one configured in git `jon.bitbucket@livewire-uk.co.uk`?

jonhoare commented:
You are right, I had signed up with my main email address jon@livewire-uk.co.uk, but I have now just signed into my linux identity account and modified the account email to be jon.bitbucket@livewire-uk.co.uk. So wonder if it has sorted itself out now?

jonhoare commented:
I've now completely resigned up and signed a new CLA with this github account and jon.bitbucket@livewire-uk.co.uk, so I'm hoping the bot will be able to pick this up now? If this doesn't work I'll have to contact the helpdesk

jonhoare commented:
It does't look like it has worked, so I have contacted the helpdesk.

jonhoare commented:
Oh, ok, forget that it does seem to have now worked. 👍

jpoon committed:
Remove Files.Get. Couldn't get that to work

jpoon commented:
Couldn't get `Files.Get` to work, so I removed it. Deployed it on a 1.6.6 cluster and everything looks to be in order: pods are up, mysql is replicating.

jpoon commented:
@viglesiasce @prydonius this PR is ready for your review.

(Thank you for your hard work in maintaining these charts)

escardin commented:
@jpoon can you expose the MYSQL_DATABASE environment variable?

escardin committed:
namespace config values into mysqlha so they can be pulled out via requirements.yaml

escardin committed:
Expose create database on deployment

escardin committed:
Oops, renamed one too many variables

escardin committed:
Update README.md

escardin commented:
CLA Signed.

thelinuxfoundation commented:
ping to bump for CLA check

unguiculus commented:
Closing as stale. Feel free to reopen the PR if/when you decide to work on it again.

jpoon commented:
@unguiculus the PR is ready for review.
@escardin did you sort out the CLA check? Seems like the bot still says you haven't signed it.

escardin commented:
I think it might be failing the CLA check because of me. @jpoon if that is the case, I'll work with you to get my commits sorted out, or you can back them out, and when your bit is merged I can make a PR for my change.

jpoon commented:
Weird, I don't have the rights to re-open this PR it looks like. 

@unguiculus is the CLA check the only thing that is blocking this from a review/merge? 
@escardin did you try emailing `helpdesk@rt.linuxfoundation.org`?

escardin commented:
Yeah, I think I need to force push some commits with the right email on them, and then they need to get into your PR somehow.

jpoon commented:
@escardin Force Push Fridays! I gave you push access to my fork (https://github.com/jpoon/charts/invitations)

escardin commented:
@jpoon I rebased, fixed my author and force pushed to your branch. Looks like the CLA is now good.
@viglesiasce I believe this is ready for re-review.

viglesiasce commented:
/ok-to-test

viglesiasce commented:
Thanks for taking this on @escardin! Worked as expected for me.
